doc/architecture/003-Implement-with-Rust.md

```diff
@@ -0,0 +1,109 @@
+
+## Context
+
+Due to significant architectural changes, NixOps 4 is a complete rewrite.
+This gives the project an opportunity to choose a different programming language.
+
+## Decision
+
+NixOps 4 will be implemented in Rust.
+
+- (+) General purpose language suitable for the purpose ("systems programming")
+- (+) Memory safety
+- (+) No garbage collector, makes integrating with a garbage-collected language easier
+- (+) Good performance
+- (+) Good Nix packaging, with multiple integrations available, which align with local development method
+- (+) Nix bindings available, although they will need to be improved, but they are mostly de-risked
+- (+) Rust has precedent in the Nix community
+  - nixpkgs-fmt, alejandra
+  - rnix-lsp, nil
+  - tvix
+  - nixpkgs-check-by-name
+  - Nix itself has experimented with a gradual Rust migration, although this was reverted because the gradual migration process was not viable
+- (-) Somewhat steep learning curve
+  - Less than Haskell's
+  - Partially mitigated by AI assistants
+
+## Alternatives
+
+### Python
+
+- (~) NixOps 1 and 2 were implemented in Python, but continuity is not relevant because NixOps 4 is a complete rewrite
+- (+) Python also has bindings available
+- (+) Python is considered to be easy to write
+  - This is less of a priority, as most development will happen in resource implementations and Nix code, which are independent of the NixOps implementation language choice.
+- (-) Python types are not as strong as Rust's.
+  - NixOps 2 showed that gradual typing + ratchet make contributing somewhat harder.
+    CI failures should be binary - not "maybe you should do a better job at typing, so let's call that a failure".
+    The alternative, no ratchet, is not acceptable, because type quality would deteriorate.
+  - Types seem to change ("improve") for no clear reason
+- (-) Python has worse performance - a risk, not necessarily a problem
+- (-) Packaging Python with Nix has proved challenging. Use of poetry2nix in Nixpkgs was discontinued, leading to a situation where the Nixpkgs packaging and local packaging would continually diverge. Python-native dependency management was not an option anymore.
+
+### Haskell
+
+- (+) Haskell has strong types
+- (+) Haskell has good performance
+- (+) Haskell has good Nix bindings (`hercules-ci-cnix-expr`), with possibly no changes needed to them
+  - (-) if well written
+- (-) Haskell has a steep learning curve
+- (-) Haskell has a its own garbage collector, making integration of the Nix language potentially riskier
+  - (+) This has proven to be a non-issue in `hercules-ci-agent`
+- (-) Haskell has a smaller community than Rust, which could make it harder to find contributors
+- (~) Haskell can be slow to iterate on, unless most iteration can happen through unit tests, which may be tricky for NixOps, of which a large part is integration testing.
+- (-) The Haskell community broadly does not have a good testing culture. (Praise to those who do)
```

Comment:
@Norfairking

Reply:
@srid This isn't an argument against Haskell, this is an argument against working with certain people who write Haskell :P

Reply:
In fact I tried starting that project in the past: https://github.com/NorfairKing/declops

Reply:
Removed in separate thread.

> Maybe not perpetuate this then 👍

- https://github.com/nixops4/nixops4/pull/5#discussion_r1587281012